lambdas/validate-oauth-callback/src/main/java/uk/gov/di/ipv/core/validateoauthcallback/ValidateOAuthCallbackHandler.java

```diff
@@ -0,0 +1,203 @@
+            AuditService auditService) {
+        this.configurationService = configurationService;
+        this.ipvSessionService = ipvSessionService;
+        this.auditService = auditService;
+    }
+
+    @ExcludeFromGeneratedCoverageReport
+    public ValidateOAuthCallbackHandler() {
+        this.configurationService = new ConfigurationService();
+        this.ipvSessionService = new IpvSessionService(configurationService);
+        this.auditService =
+                new AuditService(AuditService.getDefaultSqsClient(), configurationService);
+    }
+
+    @Override
+    @Tracing
+    @Logging(clearState = true)
+    public APIGatewayProxyResponseEvent handleRequest(
+            APIGatewayProxyRequestEvent input, Context context) {
+        LogHelper.attachComponentIdToLogs();
+
+        IpvSessionItem ipvSessionItem = null;
+        String credentialIssuerId = null;
+
+        try {
+            String ipvSessionId = RequestHelper.getIpvSessionId(input);
+
+            CredentialIssuerRequestDto request =
+                    RequestHelper.convertRequest(input, CredentialIssuerRequestDto.class);
+
+            validate(request);
+
+            ipvSessionItem = ipvSessionService.getIpvSession(ipvSessionId);
+            credentialIssuerId = ipvSessionItem.getCredentialIssuerSessionDetails().getCriId();
+
+            ClientSessionDetailsDto clientSessionDetailsDto =
+                    ipvSessionItem.getClientSessionDetails();
+            String userId = clientSessionDetailsDto.getUserId();
+
+            AuditEventUser auditEventUser =
+                    new AuditEventUser(
+                            userId,
+                            ipvSessionId,
+                            clientSessionDetailsDto.getGovukSigninJourneyId());
+            String componentId =
+                    configurationService.getSsmParameter(
+                            ConfigurationVariable.AUDIENCE_FOR_CLIENTS);
+
+            auditService.sendAuditEvent(
+                    new AuditEvent(
+                            AuditEventTypes.IPV_CRI_AUTH_RESPONSE_RECEIVED,
+                            componentId,
+                            auditEventUser));
+
+            AuthorizationCode authorizationCode =
+                    new AuthorizationCode(request.getAuthorizationCode());
+
+            setIpvSessionCRIAuthorizationCode(ipvSessionItem, authorizationCode);
+
+            ipvSessionService.updateIpvSession(ipvSessionItem);
+
+            return ApiGatewayResponseGenerator.proxyJsonResponse(
+                    HttpStatus.SC_OK, JOURNEY_NEXT_RESPONSE);
+        } catch (HttpResponseExceptionWithErrorBody e) {
+            ErrorResponse errorResponse = e.getErrorResponse();
+            LogHelper.logOauthError(
+                    "Error in credential issuer return lambda",
+                    errorResponse.getCode(),
+                    errorResponse.getMessage());
+            return ApiGatewayResponseGenerator.proxyJsonResponse(
```

Comment:
Do we need to set the visitedCredentials here as well?

Reply:
I think the rationale for not doing that here was because we're not doing so for HttpResponse errors in the retrieve-cri-oauth-access-token handler, because this error is not recoverable?